RFC: Update calling-c-and-fortran docs

kmsquire opened the pull request:
This is mostly formatting changes (e.g., many of the tables on the current page are only half-visible, at least on Chrome on OSX), but I've included a few wording updates as I've attempted to understand what the docs are referring to.  I also changed most parenthetical notes to footnotes.

This is a work in progress, but any comments/corrections/suggestions appreciated.

Cc: @vtjnash

kmsquire committed:
Update calling-c-and-fortran docs

* Formatting changes/fixes
* Small wording updates
* Removed references to #2818

stevengj commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -176,11 +172,13 @@ declared as::
+we need to write a comparison function that works for some arbitrary type
+``T``::
 
     function mycompare{T}(a::T, b::T)
```
The `T` can be dropped here; we can just write `mycompare(a, b)`

kmsquire replied:
Thanks.  Updated this example and the corresponding docs.

stevengj commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -176,11 +172,13 @@ declared as::
+``T``::
 
     function mycompare{T}(a::T, b::T)
         return convert(Cint, a < b ? -1 : a > b ? +1 : 0)::Cint
```
This can now just be `Cint(....)::Cint`.   (Is the typeassert still needed?)

kmsquire commented:
(Still making my way through the rest of the docs.)

tkelman commented:
Bump. This needs rebasing apparently.

kmsquire commented:
Rebased.  I also removed references to #2818, which was apparently moved to #10579 and merged.  @keno, @vtjnash, or @JeffBezanson, can you please look over those changes to make sure they're correct?

The docs in this area could probably use some more work, but this should be an improvement.

stevengj commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -86,7 +83,7 @@ of an environment variable, one makes a call like this::
     julia> bytestring(path)
     "/bin/bash"
 
-Note that the argument type tuple must be written as ``(Ptr{UInt8},)``,
```
Actually, in 0.4 it should be written as `(Cstring,)` since it is a pointer to NUL-terminated bytestring.

kmsquire replied:
Thanks.  I've updated the example above and following section, and will push in a second.

stevengj replied:
The return type should still be `Ptr{UInt8}`.

stevengj replied:
I take it back, since the `bytestring` method accepts a `Cstring`, it's not crazy to use a `Cstring` as a return value.  e.g. `bytestring(ccall(:strdup, Cstring, (Cstring,), "hello"))` works (although it has a memory leak).

stevengj replied:
See also #12889.

stevengj commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -215,90 +212,87 @@ it is really calling the Julia function that you passed to it).
-call signatures accurately reflect those in the C header file. (The `Clang
-package` <https://github.com/ihnorton/Clang.jl> can be used to auto-generate
-Julia code from a C header file.)
+functionsl; you are responsible for making sure that your Julia types and
```
`functionsl`?

stevengj commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -433,42 +430,63 @@ C name                  Standard Julia Alias    Julia Base Type
-`Note`: C functions that take an argument of the type ``char**`` can be called by using
-a ``Ptr{Ptr{UInt8}}`` type within Julia. For example,
-C functions of the form::
+- For ``wchar_t*`` arguments, the Julia type should be ``Ptr{Wchar_t}``, and data can
```
This seems redundant with the `For ``wchar_t*`` arguments, ...` clause below.

kmsquire replied:
Thanks, yes, I missed this one during the rebase.

stevengj commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -927,3 +942,30 @@ Chaining (parentheses optional, but recommended for readability)::
+          Base.cconvert(Int32, Base.cconvert_gcroot(Int32, x)),
+          Base.cconvert(Float64, Base.cconvert_gcroot(Float64, y)))
+
+   Note that the primary fall-back method for ``cconvert`` is::
```
inconsistent hyphenation of "fallback"

kmsquire commented:
Thanks, @stevengj.  I've fixed the parts you pointed out.

tkelman commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -215,90 +212,87 @@ it is really calling the Julia function that you passed to it).
-call signatures accurately reflect those in the C header file. (The `Clang
-package` <https://github.com/ihnorton/Clang.jl> can be used to auto-generate
-Julia code from a C header file.)
+functional; you are responsible for making sure that your Julia types and
```
`functional` looks wrong here too? autocorrect's fault?

nalimilan commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -433,42 +430,58 @@ C name                  Standard Julia Alias    Julia Base Type
-or either ``Ptr{Cchar}`` or ``Ptr{UInt8}`` otherwise (these two pointer types have the same effect), as described above,
-not ``ASCIIString``. Similarly, for array arguments (``T[]`` or ``T*``), the Julia
-type should again be ``Ptr{T}``, not ``Vector{T}``.
+- For string arguments (``char*``) the Julia type should be ``Cstring`` (if
```
I would get rid of the parentheses around the "if" here. This is an essential point, and the "otherwise" is outside of the parentheses.

nalimilan commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -433,42 +430,58 @@ C name                  Standard Julia Alias    Julia Base Type
+  Data can be converted to/from ordinary Julia strings by the ``wstring(s)``
+  function (equivalent to either ``utf16(s)`` or ``utf32(s)`` depending upon
+  the width of ``Cwchar_t``); this conversion will be called
+  automatically for ``Cwstring`` arguments.    Note also that ASCII, UTF-8,
```
Why so many spaces after the dot here?

quinnj commented:
Seems like there are some changes here we'd still want, if @kmsquire or anybody else is up for one more rebase/round of updates?

kmsquire commented:
Redone in #31700
